<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>LimROTS: A Hybrid Method Integrating Empirical Bayes and Reproducibility-Optimized Statistics for Robust Analysis of Proteomics and Metabolomics Data — LimROTS • LimROTS</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="LimROTS: A Hybrid Method Integrating Empirical Bayes and Reproducibility-Optimized Statistics for Robust Analysis of Proteomics and Metabolomics Data — LimROTS"><meta name="description" content="LimROTS: A Hybrid Method Integrating Empirical Bayes and
Reproducibility-Optimized Statistics for Robust Analysis of Proteomics and
Metabolomics Data"><meta property="og:description" content="LimROTS: A Hybrid Method Integrating Empirical Bayes and
Reproducibility-Optimized Statistics for Robust Analysis of Proteomics and
Metabolomics Data"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">LimROTS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/LimROTS.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/AliYoussef96/LimROTS/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1><code>LimROTS</code>: A Hybrid Method Integrating Empirical Bayes and Reproducibility-Optimized Statistics for Robust Analysis of Proteomics and Metabolomics Data</h1>
      <small class="dont-index">Source: <a href="https://github.com/AliYoussef96/LimROTS/blob/main/R/LimROTS.r" class="external-link"><code>R/LimROTS.r</code></a></small>
      <div class="d-none name"><code>LimROTS.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>LimROTS</code>: A Hybrid Method Integrating Empirical Bayes and
Reproducibility-Optimized Statistics for Robust Analysis of Proteomics and
Metabolomics Data</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">LimROTS</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  B <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  K <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  a1 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  a2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  log <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  progress <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  meta.info <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cluster <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  group.name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  formula.str <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  seed.cl <span class="op">=</span> <span class="fl">1234</span>,</span>
<span>  robust <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  trend <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  permutating.group <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A <code>SummarizedExperiment</code> object or a matrix where rows
represent features (e.g., proteins, metabolites) and columns
represent samples.
The values should be log-transformed.</p></dd>


<dt id="arg-b">B<a class="anchor" aria-label="anchor" href="#arg-b"></a></dt>
<dd><p>An integer representing the amount of bootstrap iterations.
Default is 1000.</p></dd>


<dt id="arg-k">K<a class="anchor" aria-label="anchor" href="#arg-k"></a></dt>
<dd><p>An optional integer representing the top list size for ranking.
If not specified, it is set to one-fourth of the number of features.</p></dd>


<dt id="arg-a-">a1<a class="anchor" aria-label="anchor" href="#arg-a-"></a></dt>
<dd><p>Optional numeric value used in the optimization process.
If defined by the user, no optimization occurs.</p></dd>


<dt id="arg-a-">a2<a class="anchor" aria-label="anchor" href="#arg-a-"></a></dt>
<dd><p>Optional numeric value used in the optimization process.
If defined by the user, no optimization occurs.</p></dd>


<dt id="arg-log">log<a class="anchor" aria-label="anchor" href="#arg-log"></a></dt>
<dd><p>Logical, indicating whether the data is already log-transformed.
Default is <code>TRUE</code>.</p></dd>


<dt id="arg-progress">progress<a class="anchor" aria-label="anchor" href="#arg-progress"></a></dt>
<dd><p>Logical, indicating whether to display a progress bar the
function's execution. Default is <code>FALSE</code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Logical, indicating whether to display messages during the
function's execution. Default is <code>TRUE</code>.</p></dd>


<dt id="arg-meta-info">meta.info<a class="anchor" aria-label="anchor" href="#arg-meta-info"></a></dt>
<dd><p>A data frame containing sample-level metadata, where each
row corresponds to a sample. It should include the grouping variable
specified in <code>group.name</code>. If <code>x</code> is a <code>SummarizedExperiment</code>
object, <code>meta.info</code> must be a vector of the metadata needed for the
model to run and can be retrieved using <code>colData()</code>.</p></dd>


<dt id="arg-cluster">cluster<a class="anchor" aria-label="anchor" href="#arg-cluster"></a></dt>
<dd><p>A parallel cluster object for distributed computation,
e.g., created by <code>makeCluster()</code>. Default is 2.</p></dd>


<dt id="arg-group-name">group.name<a class="anchor" aria-label="anchor" href="#arg-group-name"></a></dt>
<dd><p>A string specifying the column in <code>meta.info</code> that
represents the groups or conditions for comparison.</p></dd>


<dt id="arg-formula-str">formula.str<a class="anchor" aria-label="anchor" href="#arg-formula-str"></a></dt>
<dd><p>A formula string for modeling.
It should include "~ 0 + ..." to exclude the intercept from the model.
All the model parameters must be present in <code>meta.info</code>.</p></dd>


<dt id="arg-seed-cl">seed.cl<a class="anchor" aria-label="anchor" href="#arg-seed-cl"></a></dt>
<dd><p>An integer specifying the seed for randomization;
if not provided, the default is 1234.</p></dd>


<dt id="arg-robust">robust<a class="anchor" aria-label="anchor" href="#arg-robust"></a></dt>
<dd><p>indicating whether robust fitting should be used.
Default is TRUE, see eBayes.</p></dd>


<dt id="arg-trend">trend<a class="anchor" aria-label="anchor" href="#arg-trend"></a></dt>
<dd><p>indicating whether to include trend fitting in the
differential expression analysis. Default is TRUE. see eBayes.</p></dd>


<dt id="arg-permutating-group">permutating.group<a class="anchor" aria-label="anchor" href="#arg-permutating-group"></a></dt>
<dd><p>Logical, If <code>TRUE</code>, the permutation for
calculating the null distribution is performed by permuting the target group
only specified in <code>group.name</code> Preserving all the other sample
information. If <code>FALSE</code>, the entire sample information retrieved from
<code>meta.info</code> will be permuted (recommended to be set to TRUE).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>An object of class <code>"list"</code> with the following elements:</p>
<dl><dt>data</dt>
<dd><p>The original data matrix.</p></dd>

<dt>B</dt>
<dd><p>The number of bootstrap samples used.</p></dd>

<dt>d</dt>
<dd><p>The optimized statistics for each feature.</p></dd>

<dt>logfc</dt>
<dd><p>Log-fold change values between groups.</p></dd>

<dt>pvalue</dt>
<dd><p>P-values computed based on the permutation samples.</p></dd>

<dt>FDR</dt>
<dd><p>False discovery rate estimates.</p></dd>

<dt>a1</dt>
<dd><p>Optimized parameter used in differential expression ranking.</p></dd>

<dt>a2</dt>
<dd><p>Optimized parameter used in differential expression ranking.</p></dd>

<dt>k</dt>
<dd><p>Top list size used for ranking.</p></dd>

<dt>corrected.logfc</dt>
<dd><p>estimate of the log2-fold-change
corresponding to the effect corrected by the s model
see topTable.</p></dd>

<dt>q_values</dt>
<dd><p>Estimated q-values using the <code>qvalue</code> package.</p></dd>

<dt>BH.pvalue</dt>
<dd><p>Benjamini-Hochberg adjusted p-values.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The <strong>LimROTS</strong> approach initially uses
limma package functionality to simulate the intensity data of
proteins and
metabolites. A linear model is subsequently fitted using the design matrix.
Empirical Bayes variance shrinking is then implemented. To obtain the
moderated t-statistics, the adjusted standard error
\(SEpost = √(s2.post)
\times unscaled SD\) for each feature is computed, along with the regression
coefficient for each feature (indicating the impact of variations in the
experimental settings). Then, by adapting a reproducibility-optimized
technique known as <a href="https://rdrr.io/pkg/ROTS/man/ROTS.html" class="external-link">ROTS</a> to establish an optimality
based on the largest overlap of top-ranked features within group-preserving
bootstrap datasets, Finally based on the optimized parameters
\(\alpha1\) and
\(\alpha2\) this equation used to calculates the final statistics:</p>
<p>$$t_{\alpha_{(p)}} = \frac{\beta_{(p)}}
         {\alpha1 + \alpha2 \times SEpost_{(p)}}$$where
\(t_{\alpha_{(p)}}\) is the final statistics for each feature,
\(\beta_{(p)}\) is the coefficient, and \(SEpost_{(p)}\)
is the the adjusted
standard error. LimROTS generates p-values from permutation samples
using the implementation available in
qvalue package, along with internal implementation of FDR
adapted from ROTS package. Additionally, the qvalue package is used
to calculate q-values, were the proportion of true null p-values is
set to the bootstrap method pi0est. We recommend using
permutation-derived p-values and qvalues.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Ritchie, M.E., Phipson, B., Wu, D., Hu, Y., Law, C.W., Shi, W., and Smyth,
G.K. (2015). limma powers differential expression analyses for
RNA-sequencing and microarray studies. Nucleic Acids Research 43(7), e47</p>
<p>Suomi T, Seyednasrollah F, Jaakkola M, Faux T, Elo L (2017). “ROTS: An
R package for reproducibility-optimized statistical testing.
” <em>PLoS computational biology</em>, <em>13</em>(5), e1005562.
<a href="doi:10.1371/journal.pcbi.1005562">doi:10.1371/journal.pcbi.1005562</a>
<a href="https://doi.org/10.1371/journal.pcbi.1005562" class="external-link">https://doi.org/10.1371/journal.pcbi.1005562</a>,
<a href="http://www.ncbi.nlm.nih.gov/pubmed/28542205" class="external-link">http://www.ncbi.nlm.nih.gov/pubmed/28542205</a></p>
<p>Elo LL, Filen S, Lahesmaa R, Aittokallio T. Reproducibility-optimized test
statistic for ranking genes in microarray studies.
IEEE/ACM Trans Comput Biol Bioinform. 2008;5(3):423-431.
<a href="doi:10.1109/tcbb.2007.1078">doi:10.1109/tcbb.2007.1078</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Example usage:</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">500</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">100</span>, ncol <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Simulated data</span></span></span>
<span class="r-in"><span><span class="va">meta.info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, each <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">formula.str</span> <span class="op">&lt;-</span> <span class="st">"~ 0 + group"</span></span></span>
<span class="r-in"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu">LimROTS</span><span class="op">(</span><span class="va">data</span>,</span></span>
<span class="r-in"><span>    meta.info <span class="op">=</span> <span class="va">meta.info</span>, group.name <span class="op">=</span> <span class="st">"group"</span>,</span></span>
<span class="r-in"><span>    formula.str <span class="op">=</span> <span class="va">formula.str</span>, B <span class="op">=</span> <span class="fl">10</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No top list size K given, using 25</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Bootstrapping samples</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                                              |                                                                      |   0%</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No cluster found; only two cores will be used!</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Optimizing parameters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating p-values</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating FDR</span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in quantile.default(pi0, prob = 0.1):</span> missing values and NaN's not allowed if 'na.rm' is FALSE</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ali Mostafa Anwar, Leo Lahti, Akewak Jeba, Eleanor Coffey.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

