<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>LimROTS: A Hybrid Method Integrating Empirical Bayes and Reproducibility-Optimized Statistics for Robust Differential Expression Analysis — LimROTS • LimROTS</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="LimROTS: A Hybrid Method Integrating Empirical Bayes and Reproducibility-Optimized Statistics for Robust Differential Expression Analysis — LimROTS"><meta property="og:description" content="LimROTS: A Hybrid Method Integrating Empirical Bayes and
Reproducibility-Optimized Statistics for Robust Differential
Expression Analysis"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">LimROTS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/LimROTS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/AliYoussef96/LimROTS/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1><code>LimROTS</code>: A Hybrid Method Integrating Empirical Bayes and Reproducibility-Optimized Statistics for Robust Differential Expression Analysis</h1>
    <small class="dont-index">Source: <a href="https://github.com/AliYoussef96/LimROTS/blob/main/R/LimROTS.r" class="external-link"><code>R/LimROTS.r</code></a></small>
    <div class="hidden name"><code>LimROTS.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>LimROTS</code>: A Hybrid Method Integrating Empirical Bayes and
Reproducibility-Optimized Statistics for Robust Differential
Expression Analysis</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">LimROTS</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  niter <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  K <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  a1 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  a2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  log <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">meta.info</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">group.name</span>,</span>
<span>  <span class="va">formula.str</span>,</span>
<span>  robust <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  trend <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  permutating.group <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A <code>SummarizedExperiment</code> object, where rows
represent features (e.g., proteins, metabolites) and columns
represent samples.
The values should be log-transformed.</p></dd>


<dt id="arg-niter">niter<a class="anchor" aria-label="anchor" href="#arg-niter"></a></dt>
<dd><p>An integer representing the amount of bootstrap iterations.
Default is 1000.</p></dd>


<dt id="arg-k">K<a class="anchor" aria-label="anchor" href="#arg-k"></a></dt>
<dd><p>An optional integer representing the top list size for ranking.
If not specified, it is set to one-fourth of the number of features.</p></dd>


<dt id="arg-a-">a1<a class="anchor" aria-label="anchor" href="#arg-a-"></a></dt>
<dd><p>Optional numeric value used in the optimization process.
If defined by the user, no optimization occurs.</p></dd>


<dt id="arg-a-">a2<a class="anchor" aria-label="anchor" href="#arg-a-"></a></dt>
<dd><p>Optional numeric value used in the optimization process.
If defined by the user, no optimization occurs.</p></dd>


<dt id="arg-log">log<a class="anchor" aria-label="anchor" href="#arg-log"></a></dt>
<dd><p>Logical, indicating whether the data is already log-transformed.
Default is <code>TRUE</code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Logical, indicating whether to display messages during the
function's execution. Default is <code>TRUE</code>.</p></dd>


<dt id="arg-meta-info">meta.info<a class="anchor" aria-label="anchor" href="#arg-meta-info"></a></dt>
<dd><p>a character vector of the metadata needed for the
model to run and can be retrieved using <code>colData()</code>.</p></dd>


<dt id="arg-bpparam">BPPARAM<a class="anchor" aria-label="anchor" href="#arg-bpparam"></a></dt>
<dd><p>A <code>BiocParallelParam</code> object specifying the
parallelization backend (e.g., <code>MulticoreParam</code>, <code>SnowParam</code>).
The default depends on the operating system: if the user is on Windows,
<code>SnowParam(workers = 2)</code> is used; otherwise,
<code>MulticoreParam(workers = 2)</code>.</p></dd>


<dt id="arg-group-name">group.name<a class="anchor" aria-label="anchor" href="#arg-group-name"></a></dt>
<dd><p>A string specifying the column in <code>meta.info</code> that
represents the groups or conditions for comparison. group.name should be
retrieved using <code>colData()</code> as factor.</p></dd>


<dt id="arg-formula-str">formula.str<a class="anchor" aria-label="anchor" href="#arg-formula-str"></a></dt>
<dd><p>A formula string for modeling.
It should include "~ 0 + ..." to exclude the intercept from the model.
All the model parameters must be present in <code>meta.info</code>.</p></dd>


<dt id="arg-robust">robust<a class="anchor" aria-label="anchor" href="#arg-robust"></a></dt>
<dd><p>indicating whether robust fitting should be used.
Default is TRUE, see <a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a>.</p></dd>


<dt id="arg-trend">trend<a class="anchor" aria-label="anchor" href="#arg-trend"></a></dt>
<dd><p>indicating whether to include trend fitting in the
differential expression analysis. Default is TRUE. see <a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a>.</p></dd>


<dt id="arg-permutating-group">permutating.group<a class="anchor" aria-label="anchor" href="#arg-permutating-group"></a></dt>
<dd><p>Logical, If <code>TRUE</code>, the permutation for
calculating the null distribution is performed by permuting the target group
only specified in <code>group.name</code> Preserving all the other sample
information. If <code>FALSE</code>, the entire sample information retrieved from
<code>meta.info</code> will be permuted (recommended to be set to FALSE).</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>An object of class <code>"SummarizedExperiment"</code> with the
following elements:</p>
<dl><dt>data</dt>
<dd><p>The original data matrix.</p></dd>

<dt>niter</dt>
<dd><p>The number of bootstrap samples used.</p></dd>

<dt>statistics</dt>
<dd><p>The optimized statistics for each feature.</p></dd>

<dt>logfc</dt>
<dd><p>Log-fold change values between groups.</p></dd>

<dt>pvalue</dt>
<dd><p>P-values computed based on the permutation samples.</p></dd>

<dt>FDR</dt>
<dd><p>False discovery rate estimates.</p></dd>

<dt>a1</dt>
<dd><p>Optimized parameter used in differential expression ranking.</p></dd>

<dt>a2</dt>
<dd><p>Optimized parameter used in differential expression ranking.</p></dd>

<dt>k</dt>
<dd><p>Top list size used for ranking.</p></dd>

<dt>corrected.logfc</dt>
<dd><p>estimate of the log2-fold-change
corresponding to the effect corrected by the s model
see <a href="https://rdrr.io/pkg/limma/man/toptable.html" class="external-link">topTable</a>.</p></dd>

<dt>q_values</dt>
<dd><p>Estimated q-values using the <code>qvalue</code> package.</p></dd>

<dt>BH.pvalue</dt>
<dd><p>Benjamini-Hochberg adjusted p-values.</p></dd>

<dt>null.statistics</dt>
<dd><p>The optimized null statistics for each feature.</p></dd>

</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>The <strong>LimROTS</strong> approach initially uses
<span class="pkg">limma</span> package functionality to simulate the intensity data of
proteins and
metabolites. A linear model is subsequently fitted using the design matrix.
Empirical Bayes variance shrinking is then implemented. To obtain the
moderated t-statistics, the adjusted standard error
\(SE_{post} = \sqrt{s^2_{\text{post}} } \times unscaled SD\)
for each feature is computed, along with the regression
coefficient for each feature (indicating the impact of variations in the
experimental settings). Then, by adapting a reproducibility-optimized
technique known as <a href="https://rdrr.io/pkg/ROTS/man/ROTS.html" class="external-link">ROTS</a> to establish an optimality
based on the largest overlap of top-ranked features within group-preserving
bootstrap datasets, Finally based on the optimized parameters
\(\alpha1\) and
\(\alpha2\) this equation used to calculates the final statistics:</p>
<p>$$t_{\alpha_{(p)}} = \frac{\beta_{(p)}}
         {\alpha1 + \alpha2 \times SEpost_{(p)}}$$where
\(t_{\alpha_{(p)}}\) is the final statistics for each feature,
\(\beta_{(p)}\) is the coefficient, and \(SEpost_{(p)}\)
is the the adjusted
standard error. LimROTS generates p-values from permutation samples
using the implementation available in
<a href="https://rdrr.io/pkg/qvalue/man/qvalue.html" class="external-link">qvalue</a> package, along with internal implementation of FDR
adapted from ROTS package. Additionally, the qvalue package is used
to calculate q-values, were the proportion of true null p-values is
set to the bootstrap method <a href="https://rdrr.io/pkg/qvalue/man/pi0est.html" class="external-link">pi0est</a>. We recommend using
permutation-derived p-values and qvalues.</p>
<p>This function processes a dataset using parallel computation. It
leverages the <span class="pkg">BiocParallel</span> framework to distribute tasks
across multiple workers, which  can significantly reduce runtime for
large datasets.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Ritchie, M.E., Phipson, B., Wu, D., Hu, Y., Law, C.W., Shi, W., and Smyth,
G.K. (2015). limma powers differential expression analyses for
RNA-sequencing and microarray studies. Nucleic Acids Research 43(7), e47</p>
<p>Suomi T, Seyednasrollah F, Jaakkola M, Faux T, Elo L (2017). “ROTS: An
R package for reproducibility-optimized statistical testing.
” <em>PLoS computational biology</em>, <em>13</em>(5), e1005562.
<a href="doi:10.1371/journal.pcbi.1005562">doi:10.1371/journal.pcbi.1005562</a>
<a href="https://doi.org/10.1371/journal.pcbi.1005562" class="external-link">https://doi.org/10.1371/journal.pcbi.1005562</a>,
<a href="http://www.ncbi.nlm.nih.gov/pubmed/28542205" class="external-link">http://www.ncbi.nlm.nih.gov/pubmed/28542205</a></p>
<p>Elo LL, Filen S, Lahesmaa R, Aittokallio T. Reproducibility-optimized test
statistic for ranking genes in microarray studies.
IEEE/ACM Trans Comput Biol Bioinform. 2008;5(3):423-431.
<a href="doi:10.1109/tcbb.2007.1078">doi:10.1109/tcbb.2007.1078</a></p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Example usage:</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">500</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">100</span>, ncol <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Simulated data</span></span></span>
<span class="r-in"><span><span class="va">meta.info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, each <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">formula.str</span> <span class="op">&lt;-</span> <span class="st">"~ 0 + group"</span></span></span>
<span class="r-in"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu">LimROTS</span><span class="op">(</span><span class="va">data</span>,</span></span>
<span class="r-in"><span>    meta.info <span class="op">=</span> <span class="va">meta.info</span>, group.name <span class="op">=</span> <span class="st">"group"</span>,</span></span>
<span class="r-in"><span>    formula.str <span class="op">=</span> <span class="va">formula.str</span>, niter <span class="op">=</span> <span class="fl">10</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No top list size K given, using 25</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Initiating limma on bootstrapped samples</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using MulticoreParam (Unix-like OS) with two workers.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Optimizing a1 and a2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Computing p-values and FDR</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> qvalue() failed (return NULL): missing values and NaN's not allowed if 'na.rm' is FALSE</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ali Mostafa Anwar, Leo Lahti, Akewak Jeba, Eleanor Coffey.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

      </footer></div>






  </body></html>

