<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>LimROTS: A Hybrid Method Integrating Empirical Bayes and Reproducibility-Optimized Statistics for Robust Analysis of Proteomics and Metabolomics Data • LimROTS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="LimROTS: A Hybrid Method Integrating Empirical Bayes and Reproducibility-Optimized Statistics for Robust Analysis of Proteomics and Metabolomics Data">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">LimROTS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/LimROTS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/AliYoussef96/LimROTS/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>LimROTS: A Hybrid Method Integrating Empirical
Bayes and Reproducibility-Optimized Statistics for Robust Analysis of
Proteomics and Metabolomics Data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/AliYoussef96/LimROTS/blob/main/vignettes/LimROTS.Rmd" class="external-link"><code>vignettes/LimROTS.Rmd</code></a></small>
      <div class="hidden name"><code>LimROTS.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Differential expression analysis is a prevalent method utilised in
the examination of diverse biological data. The
reproducibility-optimized test statistic (ROTS) <span class="citation">Suomi et al. (2017)</span> has been developed with a
modified t-statistic based on the data’s intrinsic characteristics and
ranks features according to their statistical significance for
differential expression between two or more groups, as shown by the
f-statistic. Focusing on proteomics and metabolomics, the current ROTS
implementation cannot account for technical or biological covariates
such as MS batches or gender differences among the samples.
Consequently, we developed LimROTS, which employs a
reproducibility-optimized test statistic utilizing the limma empirical
bayes <span class="citation">Ritchie et al. (2015)</span> methodology to
simulate more complex experimental designs.</p>
</div>
<div class="section level2">
<h2 id="algorithm-overview">Algorithm overview<a class="anchor" aria-label="anchor" href="#algorithm-overview"></a>
</h2>
<p>The <strong>LimROTS</strong> approach initially uses the limma
package <span class="citation">Ritchie et al. (2015)</span> to simulate
the intensity data of proteins and metabolites. A linear model is
subsequently fitted using the design matrix. Empirical Bayes variance
shrinking is then implemented. To obtain the moderated t-statistics (or
f-statistics), the adjusted standard error
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>S</mi><msub><mi>E</mi><mrow><mi>p</mi><mi>o</mi><mi>s</mi><mi>t</mi></mrow></msub><mo>=</mo><msqrt><msubsup><mi>s</mi><mtext mathvariant="normal">post</mtext><mn>2</mn></msubsup></msqrt><mo>×</mo><mtext mathvariant="normal">unscaled SD</mtext></mrow><annotation encoding="application/x-tex">SE_{post} = \sqrt{s^2_{\text{post}}} \times \text{unscaled SD}</annotation></semantics></math>
for each feature is computed, along with the regression coefficient for
each feature (indicating the impact of variations in the experimental
settings). Then, by adapting a reproducibility-optimized technique known
as ROTS <span class="citation">Suomi et al. (2017)</span> to establish
an optimality based on the largest overlap of top-ranked features within
group-preserving bootstrap datasets (refer to <span class="citation">Elo
et al. (2008)</span> for further information on the
reproducibility-optimization). Finally based on the optimized parameters
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>α</mi><mn>1</mn></mrow><annotation encoding="application/x-tex">{\alpha1}</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>α</mi><mn>2</mn></mrow><annotation encoding="application/x-tex">{\alpha2}</annotation></semantics></math>
this equation used to calculates the final statistics:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>t</mi><mrow><mi>α</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>p</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><mo>=</mo><mfrac><msub><mi>β</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>p</mi><mo stretchy="true" form="postfix">)</mo></mrow></msub><mrow><mi>α</mi><mn>1</mn><mo>+</mo><mi>α</mi><mn>2</mn><mo>×</mo><msub><mtext mathvariant="normal">SEpost</mtext><mrow><mo stretchy="true" form="prefix">(</mo><mi>p</mi><mo stretchy="true" form="postfix">)</mo></mrow></msub></mrow></mfrac></mrow><annotation encoding="application/x-tex">t{\alpha(p)} = \frac{\beta_{(p)}}{\alpha1 + \alpha2 \times 
\text{SEpost}_{(p)}}</annotation></semantics></math></p>
<p>where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>t</mi><mrow><mi>α</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>p</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></msub><annotation encoding="application/x-tex">t_{\alpha(p)}</annotation></semantics></math>
is the final statistics for each feature,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>β</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>p</mi><mo stretchy="true" form="postfix">)</mo></mrow></msub><annotation encoding="application/x-tex">{\beta_{(p)}}</annotation></semantics></math>
is the coefficient, and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>S</mi><mi>E</mi><mi>p</mi><mi>o</mi><mi>s</mi><msub><mi>t</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>p</mi><mo stretchy="true" form="postfix">)</mo></mrow></msub></mrow><annotation encoding="application/x-tex">SEpost_{(p)}</annotation></semantics></math>
is the adjusted standard error. LimROTS generates p-values from
permutation samples using the implementation available in
<code>qvalue</code> package <span class="citation">Storey et al.
(2024)</span>, along with internal implementation of FDR adapted from
ROTS package <span class="citation">Suomi et al. (2017)</span>.
Additionally, the <code>qvalue</code> package is used to calculate
q-values, were the proportion of true null p-values is set to the
bootstrap method. We recommend using permutation-derived p-values and
qvalues.</p>
</div>
<div class="section level2">
<h2 id="computational-power">Computational Power<a class="anchor" aria-label="anchor" href="#computational-power"></a>
</h2>
<p>The number of samples, features, bootstrap iterations, and
<code>k</code>, which denotes the top list size for ranking, are the
four primary elements that determine the amount of computing resources
required for the optimisation process in LimROTS. It is therefore
advised to use at least 4 workers to execute LimROTS since it uses a
parallel processing (using <code>BiocParallel</code> <span class="citation">Morgan et al. (2024)</span>) implementation for the
bootstrapping step. The optimisation process is sequential and maybe
time-consuming, based on the <code>k</code> value; it is planned to be
modified in order to make the upcoming LimROTS version faster.</p>
</div>
<div class="section level2">
<h2 id="parameter-description-for-limrots-main-function">Parameter Description for LimROTS Main Function<a class="anchor" aria-label="anchor" href="#parameter-description-for-limrots-main-function"></a>
</h2>
<p>The <code><a href="../reference/LimROTS.html">LimROTS()</a></code> function serves as the primary interface
function for end users, providing core functionality for the analysis
implemented within the <code>LimROTS</code> package.</p>
<p><code><a href="../reference/LimROTS.html">LimROTS()</a></code>takes several parameters, and it should be
called correctly to obtain the desired output. For a detailed
description of all parameters and their usage, refer to the function’s
help page by typing <code><a href="../reference/LimROTS.html">?LimROTS</a></code> in the R console.</p>
</div>
<div class="section level2">
<h2 id="ups1-case-study">UPS1 Case Study<a class="anchor" aria-label="anchor" href="#ups1-case-study"></a>
</h2>
<p>To demonstrate LimROTS’ ability to detect true negatives complex
scenarios, we are using a DIA proteomics data from a UPS1-spiked E. coli
protein mixture <span class="citation">Gotti et al. (2022)</span>
includes 48 samples: 24 samples analyzed with Spectronaut and another 24
analyzed with ScaffoldDIA software, with total of 1656 proteins. Eight
different concentrations of UPS1 were used (0.1 to 50 fmol), grouped
into two categories: low concentrations (0.1–2.5 fmol, labeled as 2, 12
Samples from each software) and high concentrations (5–50 fmol, labeled
as 1, 12 Samples from each software).</p>
<p>A synthetic, unbalanced fake batches assigned to the samples. The
assignment follows the ratio of:</p>
<pre><code><span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;      1  2</span></span>
<span><span class="co">#&gt;   F 18  6</span></span>
<span><span class="co">#&gt;   M  6 18</span></span></code></pre>
<p>Additionally, 100 E. coli proteins were randomly selected, and an
effect size of 10 was added to each in only one of the fake batches. The
expected outcome is that only the UPS1 human proteins will be identified
as truly significant, while none of the remaining proteins should show
significant differences between the concentration groups.</p>
<p>This scenario resembles a real-world case where the experiment
involves unbalanced batch assignments or, for instance, an uneven gender
ratio among the samples.</p>
<p>LimROTS can take a SummarizedExperiment object with all the metadata
needed to run the model. In this example we importing UPS1.Case4 data
available in LimROTS.</p>
<p>The original source of the dataset can be found here <span class="citation">Gotti et al. (2022)</span></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load necessary packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/AliYoussef96/LimROTS" class="external-link">LimROTS</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the dataset</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"UPS1.Case4"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">UPS1.Case4</span><span class="op">)</span></span>
<span><span class="co">#&gt; class: SummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 1656 48 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): norm</span></span>
<span><span class="co">#&gt; rownames(1656): O00762_HUMAN O76070_HUMAN ... Q59385_ECOLI Q93K97_ECOLI</span></span>
<span><span class="co">#&gt; rowData names(2): GeneID Description</span></span>
<span><span class="co">#&gt; colnames(48): UPS1_0_1fmol_inj1_x UPS1_0_1fmol_inj2_x ...</span></span>
<span><span class="co">#&gt;   UPS1_50fmol_inj2_y UPS1_50fmol_inj3_y</span></span>
<span><span class="co">#&gt; colData names(4): SampleID tool Conc. fake.batch</span></span></code></pre></div>
<div class="section level4">
<h4 id="run-limrots">Run LimROTS<a class="anchor" aria-label="anchor" href="#run-limrots"></a>
</h4>
<p><strong>Before running LimROTS, the seed should be set for
reproducibility.</strong></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span>, kind <span class="op">=</span> <span class="st">"default"</span> , sample.kind <span class="op">=</span> <span class="st">"default"</span><span class="op">)</span></span></code></pre></div>
<p>NOTE: The target groups in the comparison should be factors 1 and 2.
For example, here, “Conc.” should have the factor levels “low” = 1 and
“high” = 2, so the “Conc.” variable should be a factor.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set metadata and formula for LimROTS analysis</span></span>
<span><span class="va">meta.info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Conc."</span>, <span class="st">"tool"</span>, <span class="st">"fake.batch"</span><span class="op">)</span></span>
<span><span class="va">niter</span> <span class="op">&lt;-</span> <span class="fl">100</span> <span class="co"># Number of bootstrap samples</span></span>
<span><span class="va">K</span> <span class="op">&lt;-</span> <span class="fl">100</span> <span class="co"># Set the value for K based on the data size</span></span>
<span><span class="va">K</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="va">K</span><span class="op">)</span></span>
<span><span class="va">group.name</span> <span class="op">&lt;-</span> <span class="st">"Conc."</span></span>
<span><span class="va">formula.str</span> <span class="op">&lt;-</span> <span class="st">"~ 0 + Conc. + tool + fake.batch"</span> <span class="co"># Formula for group comparison</span></span>
<span></span>
<span><span class="co"># Run LimROTS analysis with trend and robust settings enabled</span></span>
<span><span class="va">UPS1.Case4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LimROTS.html">LimROTS</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">UPS1.Case4</span>,</span>
<span>    niter <span class="op">=</span> <span class="va">niter</span>, K <span class="op">=</span> <span class="va">K</span>, meta.info <span class="op">=</span> <span class="va">meta.info</span>,</span>
<span>    BPPARAM  <span class="op">=</span> <span class="cn">NULL</span>, group.name <span class="op">=</span> <span class="va">group.name</span>,</span>
<span>    formula.str <span class="op">=</span> <span class="va">formula.str</span>, trend <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    robust <span class="op">=</span> <span class="cn">TRUE</span>, permutating.group <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Data is SummarizedExperiment object</span></span>
<span><span class="co">#&gt; Assay: norm will be used</span></span>
<span><span class="co">#&gt; Initiating limma on bootstrapped samples</span></span>
<span><span class="co">#&gt; Using MulticoreParam (Unix-like OS) with two workers.</span></span>
<span><span class="co">#&gt; Optimizing a1 and a2</span></span>
<span><span class="co">#&gt; Computing p-values and FDR</span></span></code></pre></div>
<p><strong>NOTE:</strong> “In this instance, we configure the number of
bootstrap iterations (niter) and the count of top-ranked features for
reproducibility optimization (K) to 100 both, in order to minimize the
example’s run-time. For actual analyses, it is advisable to utilize a
greater number of bootstraps (e.g., 1000). Also, for the number of cores
to use we recommend to use at least 4 workers</p>
<p>On Windows OS, the user can define a <code>BPPARAM</code> to work
with more than 2 workers as follows:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># BPPARAM  &lt;- SnowParam(workers = 4)</span></span>
<span><span class="co"># Commented out in the vignette to ensure the limit set by the</span></span>
<span><span class="co"># R_CHECK_LIMIT_CORES environment variable is not exceeded.</span></span></code></pre></div>
<p>On Linux and Mac OS</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># BPPARAM  &lt;- MulticoreParam(workers = 4)</span></span>
<span><span class="co"># Commented out in the vignette to ensure the limit set by the</span></span>
<span><span class="co"># R_CHECK_LIMIT_CORES environment variable is not exceeded.</span></span></code></pre></div>
<p>And then pass the <code>BPPARAM</code> to <code><a href="../reference/LimROTS.html">LimROTS()</a></code>.</p>
<p>The results from the <code>LimROTS</code> function will be appended
to the <code>SummarizedExperiment</code> object used, in this case,
<code>UPS1.Case4</code>.</p>
</div>
<div class="section level4">
<h4 id="volcano-plot-with-ggplot2">Volcano Plot with ggplot2<a class="anchor" aria-label="anchor" href="#volcano-plot-with-ggplot2"></a>
</h4>
<p>Utilising a Volcano plot and mapping the human UPS1 proteins at
q-values 5%, it is evident that LimROTS accurately identified the
majority of actual positive proteins while detecting a limited number of
simulated E.coli proteins.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a data frame from the LimROTS results</span></span>
<span></span>
<span><span class="va">limrots.result.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">UPS1.Case4</span><span class="op">)</span>, </span>
<span>                                row.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">UPS1.Case4</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Mark proteins as true positives (HUMAN UPS1 proteins)</span></span>
<span><span class="va">limrots.result.df</span><span class="op">$</span><span class="va">TP</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"HUMAN"</span>, <span class="va">limrots.result.df</span><span class="op">$</span><span class="va">GeneID</span><span class="op">)</span>,</span>
<span>    <span class="st">"HUMAN_TP"</span>, <span class="st">"ECOLI_FP"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create a volcano plot</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">limrots.result.df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">corrected.logfc</span>, y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">qvalue</span><span class="op">)</span>,</span>
<span>    color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">TP</span><span class="op">)</span></span>
<span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>        title <span class="op">=</span> <span class="st">"Volcano Plot"</span>, x <span class="op">=</span> <span class="st">"Log Fold Change"</span>, y <span class="op">=</span> <span class="st">"-Log10 q.value"</span>,</span>
<span>        color <span class="op">=</span> <span class="st">"True Positive"</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grey"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, color <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span></span></code></pre></div>
<p><img src="LimROTS_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="quality-control-plots">Quality Control Plots<a class="anchor" aria-label="anchor" href="#quality-control-plots"></a>
</h4>
<p>LimROTS generates p-values from permutation samples, along with FDR.
Additionally, the <code>qvalue</code> package is used to calculate
q-values and Benjamini-Hochberg adjusted p-values based on the
permutation-derived p-values. These can be used as Quality Control for
the LimROTS results. We recommend using permutation-derived p-values and
qvalues, though they should generally be very similar to the FDR and
Benjamini-Hochberg adjusted p-values.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Quality Control Plots</span></span>
<span></span>
<span><span class="co"># Plot of q-values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="va">UPS1.Case4</span><span class="op">)</span><span class="op">[[</span><span class="st">"q_values"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="LimROTS_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Histogram of q-values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="va">UPS1.Case4</span><span class="op">)</span><span class="op">[[</span><span class="st">"q_values"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="LimROTS_files/figure-html/unnamed-chunk-11-2.png" width="700"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Summary of q-values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu">metadata</span><span class="op">(</span><span class="va">UPS1.Case4</span><span class="op">)</span><span class="op">[[</span><span class="st">"q_values"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="comparison-of-limrots-limma-and-rots">Comparison of LimROTS, limma, and ROTS<a class="anchor" aria-label="anchor" href="#comparison-of-limrots-limma-and-rots"></a>
</h2>
<p>Before comparing the LimROTS, limma, and ROTS packages, we will need
to install the necessary packages: limma, ROTS and caret (for the use in
the comparison).</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/limma/" class="external-link">limma</a></span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"limma"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va">ROTS</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"ROTS"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://github.com/topepo/caret/" class="external-link">caret</a></span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="va">caret</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="section level3">
<h3 id="run-rots">Run ROTS<a class="anchor" aria-label="anchor" href="#run-rots"></a>
</h3>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ROTS</span><span class="op">)</span></span>
<span><span class="va">groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">UPS1.Case4</span><span class="op">$</span><span class="va">Conc.</span><span class="op">)</span></span>
<span><span class="va">rots.result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ROTS/man/ROTS.html" class="external-link">ROTS</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">UPS1.Case4</span><span class="op">)</span>, groups <span class="op">=</span> <span class="va">groups</span>, B <span class="op">=</span> <span class="va">niter</span>,</span>
<span>    K <span class="op">=</span> <span class="va">K</span>, seed <span class="op">=</span> <span class="fl">1234</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Bootstrapping samples</span></span>
<span><span class="co">#&gt; Optimizing parameters</span></span>
<span><span class="co">#&gt; Calculating p-values</span></span>
<span><span class="co">#&gt; Warning in pvalue(observed, permuted): subscript out of bounds (index 331200 &gt;=</span></span>
<span><span class="co">#&gt; vector size 331200)</span></span>
<span><span class="co">#&gt; Calculating FDR</span></span>
<span><span class="va">rots.result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    proteins <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">rots.result</span><span class="op">$</span><span class="va">data</span><span class="op">)</span>,</span>
<span>    logFC <span class="op">=</span> <span class="va">rots.result</span><span class="op">$</span><span class="va">logfc</span>, FDR <span class="op">=</span> <span class="va">rots.result</span><span class="op">$</span><span class="va">FDR</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run-limma">Run limma<a class="anchor" aria-label="anchor" href="#run-limma"></a>
</h3>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/limma/" class="external-link">limma</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'limma'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:BiocGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     plotMA</span></span>
<span></span>
<span><span class="va">design.matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/formula.html" class="external-link">formula</a></span><span class="op">(</span><span class="va">formula.str</span><span class="op">)</span>, data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">UPS1.Case4</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">lmFit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">UPS1.Case4</span><span class="op">)</span>, <span class="va">design.matrix</span><span class="op">)</span></span>
<span><span class="va">cont_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/makeContrasts.html" class="external-link">makeContrasts</a></span><span class="op">(</span><span class="st">"Conc.1-Conc.2"</span>, levels <span class="op">=</span> <span class="va">design.matrix</span><span class="op">)</span></span>
<span><span class="va">fit2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/contrasts.fit.html" class="external-link">contrasts.fit</a></span><span class="op">(</span><span class="va">fit</span>, <span class="va">cont_matrix</span><span class="op">)</span></span>
<span><span class="va">fit.ebayes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></span><span class="op">(</span><span class="va">fit2</span>, trend <span class="op">=</span> <span class="cn">TRUE</span>, robust <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">limma.result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html" class="external-link">topTable</a></span><span class="op">(</span><span class="va">fit.ebayes</span>, coef <span class="op">=</span> <span class="st">"Conc.1-Conc.2"</span>, number <span class="op">=</span> <span class="st">"Inf"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="generating-a-confusion-matrix-for-comparison">Generating a Confusion Matrix for Comparison<a class="anchor" aria-label="anchor" href="#generating-a-confusion-matrix-for-comparison"></a>
</h3>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/topepo/caret/" class="external-link">caret</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span>, warn.conflicts <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'caret'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:generics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     train</span></span>
<span></span>
<span></span>
<span><span class="va">TP</span> <span class="op">&lt;-</span> <span class="fu">elementMetadata</span><span class="op">(</span><span class="va">UPS1.Case4</span><span class="op">)</span><span class="op">[[</span><span class="st">"GeneID"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">TP</span> <span class="op">&lt;-</span> <span class="va">TP</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"HUMAN"</span>, <span class="va">TP</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span></span>
<span><span class="va">predictions_limrots</span> <span class="op">&lt;-</span> <span class="fu">elementMetadata</span><span class="op">(</span><span class="va">UPS1.Case4</span><span class="op">)</span><span class="op">[[</span><span class="st">"qvalue"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;</span> <span class="fl">0.05</span></span>
<span><span class="va">predictions_limrots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">predictions_limrots</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">true_labels_limrots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">UPS1.Case4</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">TP</span>,</span>
<span>    <span class="cn">TRUE</span>, <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">true_labels_limrots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">true_labels_limrots</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">conf_matrix_limrots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/caret/man/confusionMatrix.html" class="external-link">confusionMatrix</a></span><span class="op">(</span></span>
<span>    <span class="va">predictions_limrots</span>,</span>
<span>    <span class="va">true_labels_limrots</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">predictions_rots</span> <span class="op">&lt;-</span> <span class="va">rots.result</span><span class="op">$</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span></span>
<span><span class="va">predictions_rots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">predictions_rots</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">true_labels_rots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">rots.result</span><span class="op">$</span><span class="va">protein</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">TP</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">true_labels_rots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">true_labels_rots</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">conf_matrix_rots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/caret/man/confusionMatrix.html" class="external-link">confusionMatrix</a></span><span class="op">(</span><span class="va">predictions_rots</span>, <span class="va">true_labels_rots</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">predictions_limma</span> <span class="op">&lt;-</span> <span class="va">limma.result</span><span class="op">$</span><span class="va">adj.P.Val</span> <span class="op">&lt;</span> <span class="fl">0.05</span></span>
<span><span class="va">predictions_limma</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">predictions_limma</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">true_labels_limma</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">limma.result</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">TP</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">true_labels_limma</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">true_labels_limma</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">conf_matrix_limma</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/caret/man/confusionMatrix.html" class="external-link">confusionMatrix</a></span><span class="op">(</span><span class="va">predictions_limma</span>, <span class="va">true_labels_limma</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="summarizing-the-comparison-results">Summarizing the Comparison Results<a class="anchor" aria-label="anchor" href="#summarizing-the-comparison-results"></a>
</h3>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">extract_metrics</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">conf_matrix</span>, <span class="va">method_name</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">metrics</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="va">conf_matrix</span><span class="op">$</span><span class="va">byClass</span><span class="op">[</span><span class="st">"Sensitivity"</span><span class="op">]</span>,</span>
<span>        <span class="va">conf_matrix</span><span class="op">$</span><span class="va">byClass</span><span class="op">[</span><span class="st">"Specificity"</span><span class="op">]</span>,</span>
<span>        <span class="va">conf_matrix</span><span class="op">$</span><span class="va">byClass</span><span class="op">[</span><span class="st">"Pos Pred Value"</span><span class="op">]</span>,</span>
<span>        <span class="va">conf_matrix</span><span class="op">$</span><span class="va">byClass</span><span class="op">[</span><span class="st">"Neg Pred Value"</span><span class="op">]</span>,</span>
<span>        <span class="va">conf_matrix</span><span class="op">$</span><span class="va">byClass</span><span class="op">[</span><span class="st">"F1"</span><span class="op">]</span>,</span>
<span>        <span class="va">conf_matrix</span><span class="op">$</span><span class="va">byClass</span><span class="op">[</span><span class="st">"Balanced Accuracy"</span><span class="op">]</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>        Method <span class="op">=</span> <span class="va">method_name</span>, Metric <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">metrics</span><span class="op">)</span>,</span>
<span>        Value <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">metrics</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">metrics_limrots</span> <span class="op">&lt;-</span> <span class="fu">extract_metrics</span><span class="op">(</span><span class="va">conf_matrix_limrots</span>, <span class="st">"limROTS"</span><span class="op">)</span></span>
<span><span class="va">metrics_rots</span> <span class="op">&lt;-</span> <span class="fu">extract_metrics</span><span class="op">(</span><span class="va">conf_matrix_rots</span>, <span class="st">"ROTS"</span><span class="op">)</span></span>
<span><span class="va">metrics_limma</span> <span class="op">&lt;-</span> <span class="fu">extract_metrics</span><span class="op">(</span><span class="va">conf_matrix_limma</span>, <span class="st">"limma"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">all_metrics</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="va">metrics_limrots</span>, <span class="va">metrics_rots</span>,</span>
<span>    <span class="va">metrics_limma</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">all_metrics</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Metric</span>, y <span class="op">=</span> <span class="va">Value</span>, fill <span class="op">=</span> <span class="va">Method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span>, position <span class="op">=</span> <span class="st">"dodge"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>        title <span class="op">=</span> <span class="st">"Comparison of Performance Case Study"</span>,</span>
<span>        y <span class="op">=</span> <span class="st">"Value"</span>, x <span class="op">=</span> <span class="st">"Metric"</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>        plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span>,</span>
<span>        axis.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span>,</span>
<span>        axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span>,</span>
<span>        axis.ticks <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_line</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p><img src="LimROTS_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<p>Based on the evaluation, limROTS emerges as the most suitable method
overall when considering the balance across all performance metrics. It
achieves a strong combination of sensitivity, specificity, predictive
values, F1 score, and balanced accuracy, suggesting its ability to
reliably identify true positives and true negatives while maintaining
consistent predictive performance. In contrast, while ROTS and limma
excel in certain metrics like sensitivity, their lower specificity and
predictive values indicate limitations in broader applicability.
Therefore, limROTS appears to provide the most consistent and balanced
performance for general use cases.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.5.0 (2025-04-11)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] caret_7.0-1                 lattice_0.22-6             </span></span>
<span><span class="co">#&gt;  [3] limma_3.64.0                ROTS_2.0.0                 </span></span>
<span><span class="co">#&gt;  [5] ggplot2_3.5.2               BiocParallel_1.42.0        </span></span>
<span><span class="co">#&gt;  [7] LimROTS_1.1.0               SummarizedExperiment_1.38.1</span></span>
<span><span class="co">#&gt;  [9] Biobase_2.68.0              GenomicRanges_1.60.0       </span></span>
<span><span class="co">#&gt; [11] GenomeInfoDb_1.44.0         IRanges_2.42.0             </span></span>
<span><span class="co">#&gt; [13] S4Vectors_0.46.0            BiocGenerics_0.54.0        </span></span>
<span><span class="co">#&gt; [15] generics_0.1.4              MatrixGenerics_1.20.0      </span></span>
<span><span class="co">#&gt; [17] matrixStats_1.5.0           BiocStyle_2.36.0           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] Rdpack_2.6.4            pROC_1.18.5             rlang_1.1.6            </span></span>
<span><span class="co">#&gt;   [4] magrittr_2.0.3          e1071_1.7-16            compiler_4.5.0         </span></span>
<span><span class="co">#&gt;   [7] dir.expiry_1.16.0       systemfonts_1.2.3       vctrs_0.6.5            </span></span>
<span><span class="co">#&gt;  [10] reshape2_1.4.4          stringr_1.5.1           pkgconfig_2.0.3        </span></span>
<span><span class="co">#&gt;  [13] crayon_1.5.3            fastmap_1.2.0           XVector_0.48.0         </span></span>
<span><span class="co">#&gt;  [16] labeling_0.4.3          rmarkdown_2.29          prodlim_2025.04.28     </span></span>
<span><span class="co">#&gt;  [19] UCSC.utils_1.4.0        nloptr_2.2.1            ragg_1.4.0             </span></span>
<span><span class="co">#&gt;  [22] purrr_1.0.4             xfun_0.52               cachem_1.1.0           </span></span>
<span><span class="co">#&gt;  [25] jsonlite_2.0.0          recipes_1.3.0           DelayedArray_0.34.1    </span></span>
<span><span class="co">#&gt;  [28] parallel_4.5.0          R6_2.6.1                bslib_0.9.0            </span></span>
<span><span class="co">#&gt;  [31] stringi_1.8.7           RColorBrewer_1.1-3      parallelly_1.44.0      </span></span>
<span><span class="co">#&gt;  [34] boot_1.3-31             rpart_4.1.24            lubridate_1.9.4        </span></span>
<span><span class="co">#&gt;  [37] jquerylib_0.1.4         Rcpp_1.0.14             bookdown_0.43          </span></span>
<span><span class="co">#&gt;  [40] iterators_1.0.14        knitr_1.50              future.apply_1.11.3    </span></span>
<span><span class="co">#&gt;  [43] timechange_0.3.0        Matrix_1.7-3            splines_4.5.0          </span></span>
<span><span class="co">#&gt;  [46] nnet_7.3-20             tidyselect_1.2.1        qvalue_2.40.0          </span></span>
<span><span class="co">#&gt;  [49] abind_1.4-8             yaml_2.3.10             timeDate_4041.110      </span></span>
<span><span class="co">#&gt;  [52] codetools_0.2-20        listenv_0.9.1           tibble_3.2.1           </span></span>
<span><span class="co">#&gt;  [55] plyr_1.8.9              basilisk.utils_1.20.0   withr_3.0.2            </span></span>
<span><span class="co">#&gt;  [58] evaluate_1.0.3          future_1.49.0           desc_1.4.3             </span></span>
<span><span class="co">#&gt;  [61] survival_3.8-3          proxy_0.4-27            pillar_1.10.2          </span></span>
<span><span class="co">#&gt;  [64] BiocManager_1.30.25     filelock_1.0.3          foreach_1.5.2          </span></span>
<span><span class="co">#&gt;  [67] reformulas_0.4.1        scales_1.4.0            minqa_1.2.8            </span></span>
<span><span class="co">#&gt;  [70] globals_0.18.0          class_7.3-23            glue_1.8.0             </span></span>
<span><span class="co">#&gt;  [73] tools_4.5.0             data.table_1.17.2       lme4_1.1-37            </span></span>
<span><span class="co">#&gt;  [76] ModelMetrics_1.2.2.2    gower_1.0.2             fs_1.6.6               </span></span>
<span><span class="co">#&gt;  [79] grid_4.5.0              rbibutils_2.3           ipred_0.9-15           </span></span>
<span><span class="co">#&gt;  [82] nlme_3.1-168            GenomeInfoDbData_1.2.14 cli_3.6.5              </span></span>
<span><span class="co">#&gt;  [85] textshaping_1.0.1       S4Arrays_1.8.0          lava_1.8.1             </span></span>
<span><span class="co">#&gt;  [88] dplyr_1.1.4             gtable_0.3.6            sass_0.4.10            </span></span>
<span><span class="co">#&gt;  [91] digest_0.6.37           SparseArray_1.8.0       farver_2.1.2           </span></span>
<span><span class="co">#&gt;  [94] htmltools_0.5.8.1       pkgdown_2.1.2           lifecycle_1.0.4        </span></span>
<span><span class="co">#&gt;  [97] hardhat_1.4.1           httr_1.4.7              statmod_1.5.0          </span></span>
<span><span class="co">#&gt; [100] MASS_7.3-65</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-elo2008reproducibility" class="csl-entry">
Elo, Laura L, Sanna Filén, Riitta Lahesmaa, and Tero Aittokallio. 2008.
<span>“Reproducibility-Optimized Test Statistic for Ranking Genes in
Microarray Studies.”</span> <em>IEEE/ACM Transactions on Computational
Biology and Bioinformatics</em> 5 (3): 423–31.
</div>
<div id="ref-GOTTI2022107829" class="csl-entry">
Gotti, Clarisse, Florence Roux-Dalvai, Charles Joly-Beauparlant, Loïc
Mangnier, Mickaël Leclercq, and Arnaud Droit. 2022. <span>“DIA
Proteomics Data from a UPS1-Spiked e.coli Protein Mixture Processed with
Six Software Tools.”</span> <em>Data in Brief</em> 41: 107829.
https://doi.org/<a href="https://doi.org/10.1016/j.dib.2022.107829" class="external-link">https://doi.org/10.1016/j.dib.2022.107829</a>.
</div>
<div id="ref-BiocParallel" class="csl-entry">
Morgan, Martin, Jiefei Wang, Valerie Obenchain, Michel Lang, Ryan
Thompson, and Nitesh Turaga. 2024. <em>BiocParallel: Bioconductor
Facilities for Parallel Evaluation</em>. <a href="https://doi.org/10.18129/B9.bioc.BiocParallel" class="external-link">https://doi.org/10.18129/B9.bioc.BiocParallel</a>.
</div>
<div id="ref-limma" class="csl-entry">
Ritchie, Matthew E, Belinda Phipson, Di Wu, Yifang Hu, Charity W Law,
Wei Shi, and Gordon K Smyth. 2015. <span>“<span class="nocase">limma</span> Powers Differential Expression Analyses for
<span>RNA</span>-Sequencing and Microarray Studies.”</span> <em>Nucleic
Acids Research</em> 43 (7): e47. <a href="https://doi.org/10.1093/nar/gkv007" class="external-link">https://doi.org/10.1093/nar/gkv007</a>.
</div>
<div id="ref-qvalue" class="csl-entry">
Storey, John D., Andrew J. Bass, Alan Dabney, and David Robinson. 2024.
<em>Qvalue: Q-Value Estimation for False Discovery Rate Control</em>. <a href="https://doi.org/10.18129/B9.bioc.qvalue" class="external-link">https://doi.org/10.18129/B9.bioc.qvalue</a>.
</div>
<div id="ref-rots" class="csl-entry">
Suomi, Tomi, Fatemeh Seyednasrollah, Maria Jaakkola, Thomas Faux, and
Laura Elo. 2017. <span>“ROTS: An r Package for Reproducibility-Optimized
Statistical Testing.”</span> <em>PLoS Computational Biology</em> 13 (5):
e1005562. <a href="https://doi.org/10.1371/journal.pcbi.1005562" class="external-link">https://doi.org/10.1371/journal.pcbi.1005562</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Ali Mostafa Anwar, Leo Lahti, Akewak Jeba, Eleanor Coffey.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

      </footer>
</div>






  </body>
</html>
