<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>LimROTS: A Hybrid Method Integrating Empirical Bayes and Reproducibility-Optimized Statistics for Robust Analysis of Proteomics and Metabolomics Data • LimROTS</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="LimROTS: A Hybrid Method Integrating Empirical Bayes and Reproducibility-Optimized Statistics for Robust Analysis of Proteomics and Metabolomics Data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">LimROTS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/LimROTS.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/AliYoussef96/LimROTS/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>LimROTS: A Hybrid Method Integrating Empirical Bayes and Reproducibility-Optimized Statistics for Robust Analysis of Proteomics and Metabolomics Data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/AliYoussef96/LimROTS/blob/main/vignettes/LimROTS.Rmd" class="external-link"><code>vignettes/LimROTS.Rmd</code></a></small>
      <div class="d-none name"><code>LimROTS.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="how-limrots-work">How LimROTS Work?<a class="anchor" aria-label="anchor" href="#how-limrots-work"></a>
</h2>
<p>The <strong>LimROTS</strong> approach initially uses the limma
package <span class="citation">Ritchie et al. (2015)</span> to simulate
the intensity data of proteins and metabolites. A linear model is
subsequently fitted using the design matrix. Empirical Bayes variance
shrinking is then implemented. To obtain the moderated t-statistics (or
f-statistics), the adjusted standard error
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo stretchy="true" form="prefix">(</mo><mi>S</mi><mi>E</mi><mi>p</mi><mi>o</mi><mi>s</mi><mi>t</mi><mo>=</mo><mo>√</mo><mrow><mo stretchy="true" form="prefix">(</mo><mi>s</mi><mn>2</mn><mi>.</mi><mi>p</mi><mi>o</mi><mi>s</mi><mi>t</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>*</mo><mi>u</mi><mi>n</mi><mi>s</mi><mi>c</mi><mi>a</mi><mi>l</mi><mi>e</mi><mi>d</mi><mi>S</mi><mi>D</mi><mo stretchy="true" form="postfix">)</mo></mrow><annotation encoding="application/x-tex">(SEpost = √(s2.post) * unscaled SD)</annotation></semantics></math>
for each feature is computed, along with the regression coefficient for
each feature (indicating the impact of variations in the experimental
settings). Then, by adapting a reproducibility-optimized technique known
as ROTS <span class="citation">Suomi et al. (2017)</span> to establish
an optimality based on the largest overlap of top-ranked features within
group-preserving bootstrap datasets (refer to <span class="citation">Elo
et al. (2008)</span> for further information on the
reproducibility-optimization). Finally based on the optimized parameters
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>α</mi><mn>1</mn></mrow><annotation encoding="application/x-tex">{\alpha1}</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>α</mi><mn>2</mn></mrow><annotation encoding="application/x-tex">{\alpha2}</annotation></semantics></math>
this equation used to calculates the final statistics:
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>t</mi><mi>_</mi><mrow><mi>α</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>p</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><mo>=</mo><mfrac><msub><mi>β</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>p</mi><mo stretchy="true" form="postfix">)</mo></mrow></msub><mrow><mi>α</mi><mn>1</mn><mo>+</mo><mi>α</mi><mn>2</mn><mo>×</mo><mi>S</mi><mi>E</mi><mi>p</mi><mi>o</mi><mi>s</mi><msub><mi>t</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>p</mi><mo stretchy="true" form="postfix">)</mo></mrow></msub></mrow></mfrac></mrow><annotation encoding="application/x-tex">t\_{\alpha(p)} = \frac{\beta_{(p)}}
{\alpha1 + \alpha2 \times SEpost_{(p)}}</annotation></semantics></math></p>
<p>where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>t</mi><mrow><mi>α</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>p</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow></msub><annotation encoding="application/x-tex">t_{\alpha(p)}</annotation></semantics></math>
is the final statistics for each feature,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>β</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>p</mi><mo stretchy="true" form="postfix">)</mo></mrow></msub><annotation encoding="application/x-tex">{\beta_{(p)}}</annotation></semantics></math>
is the coefficient, and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>S</mi><mi>E</mi><mi>p</mi><mi>o</mi><mi>s</mi><msub><mi>t</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>p</mi><mo stretchy="true" form="postfix">)</mo></mrow></msub></mrow><annotation encoding="application/x-tex">SEpost_{(p)}</annotation></semantics></math>
is the adjusted standard error. LimROTS generates p-values from
permutation samples using the implementation available in
<code>qvalue</code> package <span class="citation">Storey et al.
(2024)</span>, along with internal implementation of FDR adapted from
ROTS package <span class="citation">Suomi et al. (2017)</span>.
Additionally, the <code>qvalue</code> package is used to calculate
q-values, were the proportion of true null p-values is set to the
bootstrap method. We recommend using permutation-derived p-values and
qvalues.</p>
</div>
<div class="section level2">
<h2 id="computational-power">Computational Power<a class="anchor" aria-label="anchor" href="#computational-power"></a>
</h2>
<p>The number of samples, features, bootstrap iterations, and
<code>k</code>, which denotes the top list size for ranking, are the
four primary elements that determine the amount of computing resources
required for the optimisation process in LimROTS. It is therefore
advised to use at least 4 cores to execute LimROTS since it uses a
parallel processing implementation for the bootstrapping step. The
optimisation process is sequential and maybe time-consuming, based on
the <code>k</code> value; it is planned to be modified in order to make
the upcoming LimROTS version faster.</p>
</div>
<div class="section level2">
<h2 id="parameter-description-for-limrots">Parameter Description for <code>LimROTS</code><a class="anchor" aria-label="anchor" href="#parameter-description-for-limrots"></a>
</h2>
<p><code>LimROTS</code> takes several parameters, and it should be
called correctly to obtain the desired output.</p>
<ul>
<li><p><strong><code>x</code></strong>: The input data, which can be a
<code>SummarizedExperiment</code> object or a matrix where rows
represent features (e.g., genes, proteins) and columns represent
samples. The values should be log-transformed.</p></li>
<li><p><strong><code>B</code></strong>: An integer representing the
amount of bootstrap iterations. Default 1000.</p></li>
<li><p><strong><code>K</code></strong>: An optional integer representing
the top list size for ranking. If not specified, it is set to one-fourth
of the number of features.</p></li>
<li><p><strong><code>a1</code></strong>: Optional numeric value used in
the optimization process. If defined by the user, no optimization
occurs.</p></li>
<li><p><strong><code>a2</code></strong>: Optional numeric value used in
the optimization process. If defined by the user, no optimization occurs
[0,1].</p></li>
<li><p><strong><code>log</code></strong>: Logical, indicating whether
the data is already log-transformed. Default is
<code>TRUE</code>.</p></li>
<li><p><strong><code>progress</code></strong>: Logical, indicating
whether to display a progress bar during bootstrap sampling. Default is
<code>FALSE</code>.</p></li>
<li><p><strong><code>verbose</code></strong>: Logical, indicating
whether to display messages during the function’s execution. Default is
<code>TRUE</code>.</p></li>
<li><p><strong><code>meta.info</code></strong>: A data frame containing
sample-level metadata, where each row corresponds to a sample. It should
include the grouping variable specified in <code>group.name</code>. If
<code>x</code> (data) is a <code>SummarizedExperiment</code> object,
<code>meta.info</code> must be a vector of the metadata needed for the
model to run and can be retrieved using <code>colData()</code>.</p></li>
<li><p><strong><code>group.name</code></strong>: A string specifying the
column in <code>meta.info</code> that represents the groups or
conditions for comparison.</p></li>
<li><p><strong><code>seed.cl</code></strong>: An integer specifying the
seed for randomization; if not provided, the default is 1234.</p></li>
<li><p><strong><code>cluster</code></strong>: A parallel cluster object
for distributed computation, e.g., created by
<code><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">makeCluster()</a></code>. Default is 2.</p></li>
<li><p><strong><code>formula.str</code></strong>: A formula string for
modeling. It should include “~ 0 + …” to exclude the intercept from the
model. All the model parameters must be present in
<code>meta.info.</code></p></li>
<li><p><strong><code>robust</code></strong>: Logical, indicating whether
robust fitting should be used. Default is <code>TRUE</code>.</p></li>
<li><p><strong><code>trend</code></strong>: Logical, indicating whether
to include trend fitting in the differential expression analysis.
Default is <code>TRUE</code>.</p></li>
<li><p><strong><code>permutating.group</code></strong>: Logical, If
<code>TRUE</code>, the permutation for calculating the null distribution
is performed by permuting the target group only specified in
<code>group.name</code> Preserving all the other sample information. If
<code>FALSE</code>, the entire sample information retrieved from
<code>meta.info</code> will be permuted (recommended to be set to
TRUE).</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="ups1-case-study">UPS1 Case Study<a class="anchor" aria-label="anchor" href="#ups1-case-study"></a>
</h2>
<p>To demonstrate LimROTS’ ability to detect true negatives complex
scenarios, we are using a DIA proteomics data from a UPS1-spiked E. coli
protein mixture <span class="citation">Gotti et al. (2022)</span>
includes 48 samples: 24 samples analyzed with Spectronaut and another 24
analyzed with ScaffoldDIA software, with total of 1656 proteins. Eight
different concentrations of UPS1 were used (0.1 to 50 fmol), grouped
into two categories: low concentrations (0.1–2.5 fmol, labeled as Conc.
2, 12 Samples from each software) and high concentrations (5–50 fmol,
labeled as Conc. 1, 12 Samples from each software).</p>
<p>A synthetic, unbalanced fake batches assigned to the samples. The
assignment follows the ratio of:</p>
<pre><code><span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;      1  2</span></span>
<span><span class="co">#&gt;   F 18  6</span></span>
<span><span class="co">#&gt;   M  6 18</span></span></code></pre>
<p>Additionally, 100 E. coli proteins were randomly selected, and an
effect size of 10 was added to each. The expected outcome is that only
the UPS1 human proteins will be identified as truly significant, while
none of the remaining proteins should show significant differences
between the concentration groups.</p>
<p>This scenario resembles a real-world case where the experiment
involves unbalanced batch assignments or, for instance, an uneven gender
ratio among the samples.</p>
<p>LimROTS can take a SummarizedExperiment object with all the metadata
needed to run the model. In this example we importing UPS1.Case4 data
available in LimROTS.</p>
<p>The original source of the dataset can be found here <span class="citation">Gotti et al. (2022)</span></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load necessary packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/AliYoussef96/LimROTS" class="external-link">LimROTS</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">parallel</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the dataset</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"UPS1.Case4"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">UPS1.Case4</span><span class="op">)</span></span>
<span><span class="co">#&gt; class: SummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 1656 48 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): norm</span></span>
<span><span class="co">#&gt; rownames(1656): O00762_HUMAN O76070_HUMAN ... Q59385_ECOLI Q93K97_ECOLI</span></span>
<span><span class="co">#&gt; rowData names(2): GeneID Description</span></span>
<span><span class="co">#&gt; colnames(48): UPS1_0_1fmol_inj1_x UPS1_0_1fmol_inj2_x ...</span></span>
<span><span class="co">#&gt;   UPS1_50fmol_inj2_y UPS1_50fmol_inj3_y</span></span>
<span><span class="co">#&gt; colData names(4): SampleID tool Conc. fake.batch</span></span></code></pre></div>
<div class="section level4">
<h4 id="run-limrots">Run LimROTS<a class="anchor" aria-label="anchor" href="#run-limrots"></a>
</h4>
<p>Although <code>seed.cl</code> internally control the seed
randomization during the parallel bootstrapping process. It is advisable
to establish a consistent seed prior to executing
<code>LimROTS</code></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span>, kind <span class="op">=</span> <span class="st">"default"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set metadata and formula for LimROTS analysis</span></span>
<span><span class="va">meta.info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Conc."</span>, <span class="st">"tool"</span>, <span class="st">"fake.batch"</span><span class="op">)</span></span>
<span><span class="va">B</span> <span class="op">&lt;-</span> <span class="fl">100</span> <span class="co"># Number of bootstrap samples</span></span>
<span><span class="va">K</span> <span class="op">&lt;-</span> <span class="fl">100</span> <span class="co"># Set the value for K based on the data size</span></span>
<span><span class="va">K</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="va">K</span><span class="op">)</span></span>
<span><span class="va">num_cores</span> <span class="op">&lt;-</span> <span class="fl">2</span> <span class="co"># Number of cores for parallel processing</span></span>
<span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/parallel/makeCluster.html" class="external-link">makeCluster</a></span><span class="op">(</span><span class="va">num_cores</span><span class="op">)</span> <span class="co"># Create a parallel cluster</span></span>
<span><span class="va">group.name</span> <span class="op">&lt;-</span> <span class="st">"Conc."</span></span>
<span><span class="va">formula.str</span> <span class="op">&lt;-</span> <span class="st">"~ 0 + Conc. + tool + fake.batch"</span> <span class="co"># Formula for group comparison</span></span>
<span></span>
<span><span class="co"># Run LimROTS analysis with trend and robust settings enabled</span></span>
<span><span class="va">limrots.result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LimROTS.html">LimROTS</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">UPS1.Case4</span>,</span>
<span>    B <span class="op">=</span> <span class="va">B</span>, K <span class="op">=</span> <span class="va">K</span>, meta.info <span class="op">=</span> <span class="va">meta.info</span>,</span>
<span>    cluster <span class="op">=</span> <span class="va">cluster</span>, group.name <span class="op">=</span> <span class="va">group.name</span>,</span>
<span>    formula.str <span class="op">=</span> <span class="va">formula.str</span>, trend <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>    robust <span class="op">=</span> <span class="cn">TRUE</span>, permutating.group <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    seed.cl <span class="op">=</span> <span class="fl">1234</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Data is SummarizedExperiment object</span></span>
<span><span class="co">#&gt; Assay: norm will be used</span></span>
<span><span class="co">#&gt; Bootstrapping samples</span></span>
<span><span class="co">#&gt;   |                                                                              |                                                                      |   0%</span></span>
<span><span class="co">#&gt; Optimizing parameters</span></span>
<span><span class="co">#&gt; Warning: One very small variance detected, has been offset away from zero</span></span>
<span><span class="co">#&gt; Calculating p-values</span></span>
<span><span class="co">#&gt; Calculating FDR</span></span></code></pre></div>
<p><strong>NOTE:</strong> “In this instance, we configure the number of
bootstrap iterations (B) and the count of top-ranked features for
reproducibility optimization (K) to 100 both, in order to minimize the
example’s run-time. For actual analyses, it is advisable to utilize a
greater number of bootstraps (e.g., 1000). Also, for the number of cores
to use we recommend to use at least 4 cores.</p>
</div>
<div class="section level4">
<h4 id="volcano-plot-with-ggplot2">Volcano Plot with ggplot2<a class="anchor" aria-label="anchor" href="#volcano-plot-with-ggplot2"></a>
</h4>
<p>Utilising a Volcano plot and mapping the human UPS1 proteins at
q-values of 1% and 5%, it is evident that LimROTS accurately identified
the majority of actual positive proteins while detecting a limited
number of simulated E.coli proteins.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a data frame from the LimROTS results</span></span>
<span><span class="va">limrots.result.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    proteins <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">limrots.result</span><span class="op">$</span><span class="va">data</span><span class="op">)</span>,</span>
<span>    LimROTS.FC <span class="op">=</span> <span class="va">limrots.result</span><span class="op">$</span><span class="va">corrected.logf</span>,</span>
<span>    q.value <span class="op">=</span> <span class="va">limrots.result</span><span class="op">$</span><span class="va">q_values</span><span class="op">$</span><span class="va">qvalues</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Mark proteins as true positives (HUMAN UPS1 proteins)</span></span>
<span><span class="va">limrots.result.df</span><span class="op">$</span><span class="va">TP</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"HUMAN"</span>, <span class="va">limrots.result.df</span><span class="op">$</span><span class="va">proteins</span><span class="op">)</span>,</span>
<span>    <span class="st">"HUMAN_TP"</span>, <span class="st">"ECOLI_FP"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create a volcano plot</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">limrots.result.df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="va">LimROTS.FC</span>, y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">q.value</span><span class="op">)</span>,</span>
<span>    color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">TP</span><span class="op">)</span></span>
<span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>        title <span class="op">=</span> <span class="st">"Volcano Plot"</span>, x <span class="op">=</span> <span class="st">"Log Fold Change"</span>, y <span class="op">=</span> <span class="st">"-Log10 q.value"</span>,</span>
<span>        color <span class="op">=</span> <span class="st">"True Positive"</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grey"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, color <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fl">0.01</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span></span></code></pre></div>
<p><img src="LimROTS_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="quality-control-plots">Quality Control Plots<a class="anchor" aria-label="anchor" href="#quality-control-plots"></a>
</h4>
<p>LimROTS generates p-values from permutation samples, along with FDR.
Additionally, the <code>qvalue</code> package is used to calculate
q-values and Benjamini-Hochberg adjusted p-values based on the
permutation-derived p-values. These can be used as Quality Control for
the LimROTS results. We recommend using permutation-derived p-values and
qvalues, though they should generally be very similar to the FDR and
Benjamini-Hochberg adjusted p-values.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Quality Control Plots</span></span>
<span></span>
<span><span class="co"># Plot of q-values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">limrots.result</span><span class="op">$</span><span class="va">q_values</span>,</span>
<span>    main <span class="op">=</span> <span class="st">"Q-values"</span>, xlab <span class="op">=</span> <span class="st">"Index"</span>,</span>
<span>    ylab <span class="op">=</span> <span class="st">"Q-value"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="LimROTS_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Histogram of q-values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">limrots.result</span><span class="op">$</span><span class="va">q_values</span>,</span>
<span>    main <span class="op">=</span> <span class="st">"Q-value Distribution"</span>,</span>
<span>    xlab <span class="op">=</span> <span class="st">"Q-value"</span>, col <span class="op">=</span> <span class="st">"lightgreen"</span>, border <span class="op">=</span> <span class="st">"white"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="LimROTS_files/figure-html/unnamed-chunk-8-2.png" width="700"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Summary of q-values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">limrots.result</span><span class="op">$</span><span class="va">q_values</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.5 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] parallel  stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] ggplot2_3.5.1    LimROTS_0.99.0   BiocStyle_2.32.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] SummarizedExperiment_1.34.0 gtable_0.3.5               </span></span>
<span><span class="co">#&gt;  [3] xfun_0.48                   bslib_0.8.0                </span></span>
<span><span class="co">#&gt;  [5] Biobase_2.64.0              lattice_0.22-6             </span></span>
<span><span class="co">#&gt;  [7] vctrs_0.6.5                 tools_4.4.1                </span></span>
<span><span class="co">#&gt;  [9] generics_0.1.3              stats4_4.4.1               </span></span>
<span><span class="co">#&gt; [11] tibble_3.2.1                fansi_1.0.6                </span></span>
<span><span class="co">#&gt; [13] highr_0.11                  pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt; [15] Matrix_1.7-0                desc_1.4.3                 </span></span>
<span><span class="co">#&gt; [17] S4Vectors_0.42.1            rngtools_1.5.2             </span></span>
<span><span class="co">#&gt; [19] lifecycle_1.0.4             GenomeInfoDbData_1.2.12    </span></span>
<span><span class="co">#&gt; [21] farver_2.1.2                stringr_1.5.1              </span></span>
<span><span class="co">#&gt; [23] compiler_4.4.1              textshaping_0.4.0          </span></span>
<span><span class="co">#&gt; [25] munsell_0.5.1               statmod_1.5.0              </span></span>
<span><span class="co">#&gt; [27] codetools_0.2-20            GenomeInfoDb_1.40.1        </span></span>
<span><span class="co">#&gt; [29] htmltools_0.5.8.1           sass_0.4.9                 </span></span>
<span><span class="co">#&gt; [31] yaml_2.3.10                 pkgdown_2.1.1              </span></span>
<span><span class="co">#&gt; [33] pillar_1.9.0                crayon_1.5.3               </span></span>
<span><span class="co">#&gt; [35] jquerylib_0.1.4             DelayedArray_0.30.1        </span></span>
<span><span class="co">#&gt; [37] cachem_1.1.0                limma_3.60.6               </span></span>
<span><span class="co">#&gt; [39] doRNG_1.8.6                 iterators_1.0.14           </span></span>
<span><span class="co">#&gt; [41] abind_1.4-8                 foreach_1.5.2              </span></span>
<span><span class="co">#&gt; [43] tidyselect_1.2.1            digest_0.6.37              </span></span>
<span><span class="co">#&gt; [45] stringi_1.8.4               reshape2_1.4.4             </span></span>
<span><span class="co">#&gt; [47] dplyr_1.1.4                 bookdown_0.40              </span></span>
<span><span class="co">#&gt; [49] labeling_0.4.3              splines_4.4.1              </span></span>
<span><span class="co">#&gt; [51] fastmap_1.2.0               grid_4.4.1                 </span></span>
<span><span class="co">#&gt; [53] colorspace_2.1-1            cli_3.6.3                  </span></span>
<span><span class="co">#&gt; [55] SparseArray_1.4.8           magrittr_2.0.3             </span></span>
<span><span class="co">#&gt; [57] S4Arrays_1.4.1              utf8_1.2.4                 </span></span>
<span><span class="co">#&gt; [59] withr_3.0.1                 scales_1.3.0               </span></span>
<span><span class="co">#&gt; [61] UCSC.utils_1.0.0            rmarkdown_2.28             </span></span>
<span><span class="co">#&gt; [63] XVector_0.44.0              httr_1.4.7                 </span></span>
<span><span class="co">#&gt; [65] matrixStats_1.4.1           qvalue_2.36.0              </span></span>
<span><span class="co">#&gt; [67] ragg_1.3.3                  evaluate_1.0.0             </span></span>
<span><span class="co">#&gt; [69] knitr_1.48                  GenomicRanges_1.56.1       </span></span>
<span><span class="co">#&gt; [71] IRanges_2.38.1              doParallel_1.0.17          </span></span>
<span><span class="co">#&gt; [73] rlang_1.1.4                 Rcpp_1.0.13                </span></span>
<span><span class="co">#&gt; [75] glue_1.8.0                  BiocManager_1.30.25        </span></span>
<span><span class="co">#&gt; [77] BiocGenerics_0.50.0         jsonlite_1.8.9             </span></span>
<span><span class="co">#&gt; [79] plyr_1.8.9                  R6_2.5.1                   </span></span>
<span><span class="co">#&gt; [81] MatrixGenerics_1.16.0       systemfonts_1.1.0          </span></span>
<span><span class="co">#&gt; [83] fs_1.6.4                    zlibbioc_1.50.0</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-elo2008reproducibility" class="csl-entry">
Elo, Laura L, Sanna Filén, Riitta Lahesmaa, and Tero Aittokallio. 2008.
<span>“Reproducibility-Optimized Test Statistic for Ranking Genes in
Microarray Studies.”</span> <em>IEEE/ACM Transactions on Computational
Biology and Bioinformatics</em> 5 (3): 423–31.
</div>
<div id="ref-GOTTI2022107829" class="csl-entry">
Gotti, Clarisse, Florence Roux-Dalvai, Charles Joly-Beauparlant, Loïc
Mangnier, Mickaël Leclercq, and Arnaud Droit. 2022. <span>“DIA
Proteomics Data from a UPS1-Spiked e.coli Protein Mixture Processed with
Six Software Tools.”</span> <em>Data in Brief</em> 41: 107829.
https://doi.org/<a href="https://doi.org/10.1016/j.dib.2022.107829" class="external-link">https://doi.org/10.1016/j.dib.2022.107829</a>.
</div>
<div id="ref-limma" class="csl-entry">
Ritchie, Matthew E, Belinda Phipson, Di Wu, Yifang Hu, Charity W Law,
Wei Shi, and Gordon K Smyth. 2015. <span>“<span class="nocase">limma</span> Powers Differential Expression Analyses for
<span>RNA</span>-Sequencing and Microarray Studies.”</span> <em>Nucleic
Acids Research</em> 43 (7): e47. <a href="https://doi.org/10.1093/nar/gkv007" class="external-link">https://doi.org/10.1093/nar/gkv007</a>.
</div>
<div id="ref-qvalue" class="csl-entry">
Storey, John D., Andrew J. Bass, Alan Dabney, and David Robinson. 2024.
<em>Qvalue: Q-Value Estimation for False Discovery Rate Control</em>. <a href="https://doi.org/10.18129/B9.bioc.qvalue" class="external-link">https://doi.org/10.18129/B9.bioc.qvalue</a>.
</div>
<div id="ref-rots" class="csl-entry">
Suomi, Tomi, Fatemeh Seyednasrollah, Maria Jaakkola, Thomas Faux, and
Laura Elo. 2017. <span>“ROTS: An r Package for Reproducibility-Optimized
Statistical Testing.”</span> <em>PLoS Computational Biology</em> 13 (5):
e1005562. <a href="https://doi.org/10.1371/journal.pcbi.1005562" class="external-link">https://doi.org/10.1371/journal.pcbi.1005562</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ali Mostafa Anwar, Leo Lahti, Akewak Jeba, Eleanor Coffey.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
